File: runtime/src/de/fatox/meta/camera/ArcCamControl.java
package de.fatox.meta.camera;

import com.badlogic.gdx.Input.Buttons;
import com.badlogic.gdx.Input.Keys;
import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.PerspectiveCamera;
import com.badlogic.gdx.graphics.VertexAttributes;
import com.badlogic.gdx.graphics.g3d.Material;
import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.attributes.ColorAttribute;
import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder;
import com.badlogic.gdx.math.Vector3;
import de.fatox.meta.Meta;
import de.fatox.meta.api.entity.EntityManager;
import de.fatox.meta.entity.Meta3DEntity;
import de.fatox.meta.injection.Inject;

/**
 * Better camera control because I dislike LibGDX's ones
 * Very similar to World Editor.
 *
 * @author Frotty
 */
public class ArcCamControl implements InputProcessor {
    private static Vector3 temp = new Vector3();
    private final Meta3DEntity targetDebug;
    /**
     * The button for moving the target.
     */
    public int moveCameraButton = Buttons.RIGHT;
    public int resetCameraButton = Buttons.MIDDLE;
    /**
     * The units to translate the camera when moved the full width or height of the screen.
     */
    public float translateUnits = 2f; // FIXME auto calculate this based on the target
    /**
     * The key which must be pressed to enter rotation mode.
     */
    public int rotateMode = Keys.CONTROL_LEFT;
    protected boolean rotateModeOn = false;
    protected boolean fastZoomMode = false;
    /**
     * The camera.
     */
    @Inject
    public PerspectiveCamera camera;
    @Inject
    public EntityManager<Meta3DEntity> entityManager;
    @Inject
    public ModelBuilder modelBuilder;
    /**
     * Are we in moveMode?
     */
    private boolean moveModeOn = false;
    /**
     * The target of the arcball
     */
    private Vector3 target = Vector3.Zero;
    /**
     * The planar (X/Y) rotation of the camera
     */
    private float rotationAngle = 0;
    /**
     * The angle in which the camera looks onto the target
     */
    private float angleOfAttack = 56;

    /**
     * Distance from target
     */
    private float distance = 50;

    public float getDistance() {
        return distance;
    }

    public void setDistance(float distance) {
        this.distance = distance;
        update();
    }

    private int startX, startY;

    public ArcCamControl() {
        Meta.inject(this);

        final Material material = new Material(ColorAttribute.createDiffuse(Color.SLATE));
        final long attributes = VertexAttributes.Usage.Position | VertexAttributes.Usage.Normal | VertexAttributes.Usage.ColorUnpacked;
        Model box = modelBuilder.createSphere(4, 4, 4, 42,42, material, attributes);
        targetDebug = new Meta3DEntity(target, box);
        update();
        entityManager.addEntity(targetDebug);
    }

    public void update() {
        target.lerp(temp, 0.5f);
        targetDebug.setPosition(target);

        camera.position.x = ppX(target.x, distance, rotationAngle, angleOfAttack);
        camera.position.y = ppY(0, distance, angleOfAttack);
        camera.position.z = ppZ(target.z, distance, rotationAngle, angleOfAttack);

        camera.direction.x = target.x - camera.position.x;
        camera.direction.y = target.y - camera.position.y;
        camera.direction.z = target.z - camera.position.z;

        camera.up.x = -sin(rotationAngle) * cos(angleOfAttack);
        camera.up.y = sin(angleOfAttack);
        camera.up.z = -cos(rotationAngle) * cos(angleOfAttack);
        camera.update();
    }

    private static float cos(float aoa) {
        return (float) Math.cos(aoa * DEGTORAD);
    }

    private static float sin(float ang) {
        return (float) Math.sin(ang * DEGTORAD);
    }

    @Override
    public boolean touchDown(int screenX, int screenY, int pointer, int button) {
        if (button == moveCameraButton) {
            startX = screenX;
            startY = screenY;
            moveModeOn = true;
        }
        if (button == resetCameraButton) {
            rotationAngle = 0;
            angleOfAttack = 56;
            distance = 50;
        }
        update();
        return false;
    }

    @Override
    public boolean touchUp(int screenX, int screenY, int pointer, int button) {
        if (button == moveCameraButton) {
            startX = screenX;
            startY = screenY;
            moveModeOn = false;
        }
        update();
        return false;
    }

    @Override
    public boolean touchDragged(int screenX, int screenY, int pointer) {
        // Touch drag equals clicking a mouseButton and then moving the mouse
        // In case the right mouse button is clicked, we are in MoveMode
        if (moveModeOn) {
            // Calculate the middle of old and new mousePosition
            float deltaX = (screenX - startX);
            float deltaY = (startY - screenY);
            startX = screenX;
            startY = screenY;
            // If CTRL is active, we only rotate
            if (rotateModeOn) {
                angleOfAttack += deltaY * .25;
                rotationAngle += deltaX * .25;
            } else {
                // Otherwise we simple move the target
                if (distance < 150) {
                    deltaX *= distance / 100;
                    deltaY *= distance / 100;
                }
                temp.set(target).add(sin(rotationAngle) * deltaY + cos(rotationAngle) * -deltaX, 0, cos(rotationAngle) * deltaY + sin(rotationAngle) * deltaX);
            }
            update();
        }
        return false;
    }

    @Override
    public boolean scrolled(int amount) {
        if (fastZoomMode) {
            return zoom(amount * translateUnits * 10);
        } else {
            return zoom(amount * translateUnits);
        }

    }

    public boolean zoom(float amount) {
        distance += amount;
        update();
        return true;
    }

    @Override
    public boolean keyDown(int keycode) {
        if (keycode == Keys.CONTROL_LEFT) {
            rotateModeOn = true;
        } else if (keycode == Keys.SHIFT_LEFT) {
            fastZoomMode = true;
        }
        return false;
    }

    @Override
    public boolean keyUp(int keycode) {
        if (keycode == Keys.CONTROL_LEFT) {
            rotateModeOn = false;
        } else if (keycode == Keys.SHIFT_LEFT) {
            fastZoomMode = false;
        }
        return false;
    }

    @Override
    public boolean keyTyped(char character) {
        return false;
    }

    @Override
    public boolean mouseMoved(int screenX, int screenY) {
        return false;
    }

    /**
     * Polar Projection from Wurst
     */
    private static float ppX(float x, float dist, float ang, float aoa) {
        return (float) (x + dist * Math.sin(ang * DEGTORAD) * Math.sin(aoa * DEGTORAD));
    }

    private static float ppY(float y, float dist, float ang) {
        return (float) (y + dist * Math.cos(ang * DEGTORAD));
    }

    private static float ppZ(float z, float dist, float ang, float aoa) {
        return (float) (z + dist * Math.cos(ang * DEGTORAD) * Math.sin(aoa * DEGTORAD));
    }

    public static final float DEGTORAD = 0.017453293f;
}

File: runtime/src/main/kotlin/de/fatox/meta/camera/ArcCamControl.kt
package de.fatox.meta.camera

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.Input.Buttons
import com.badlogic.gdx.InputProcessor
import com.badlogic.gdx.graphics.PerspectiveCamera
import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder
import com.badlogic.gdx.math.Vector3
import de.fatox.meta.api.entity.EntityManager
import de.fatox.meta.entity.Meta3DEntity
import de.fatox.meta.injection.MetaInject.Companion.lazyInject
import de.fatox.meta.sound.dlerp

/** Arcball-style editor camera control with pan, rotate, reset, and wheel zoom gestures. */
class ArcCamControl : InputProcessor {
	/** Pointer button used to pan or rotate around the target. */
	var moveCameraButton: Int = Buttons.RIGHT
	var resetCameraButton: Int = Buttons.MIDDLE

	/** Distance adjustment per wheel unit. Holding Shift multiplies it by ten. */
	var translateUnits: Float = 0.2f

	/** Key held while dragging to rotate instead of pan. */
	var rotateMode: Int = Input.Keys.CONTROL_LEFT
	private var rotateModeOn: Boolean = false
	private var fastZoomMode: Boolean = false

	val camera: PerspectiveCamera by lazyInject()
	val entityManager: EntityManager<Meta3DEntity> by lazyInject()
	val modelBuilder: ModelBuilder by lazyInject()

	private var moveModeOn = false

	private val target = Vector3()

	private var rotationAngle = 0f

	private var angleOfAttack = 56f

	private var distance = 10f
	fun getDistance(): Float {
		return distance
	}

	fun setDistance(distance: Float) {
		this.distance = distance
		update(Gdx.graphics.deltaTime)
	}

	private var startX = 0
	private var startY = 0
	fun update(delta: Float) {
		target.dlerp(temp, 0.5f, delta)
		camera.position.x = ppX(target.x, distance, rotationAngle, angleOfAttack)
		camera.position.y = ppY(0f, distance, angleOfAttack)
		camera.position.z = ppZ(target.z, distance, rotationAngle, angleOfAttack)
		camera.direction.set(target.x - camera.position.x, target.y - camera.position.y, target.z - camera.position.z).nor()
		camera.up.x = -sin(rotationAngle) * cos(angleOfAttack)
		camera.up.y = sin(angleOfAttack)
		camera.up.z = -cos(rotationAngle) * cos(angleOfAttack)
		camera.update()
	}

	override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
		if (button == moveCameraButton) {
			startX = screenX
			startY = screenY
			moveModeOn = true
		}
		if (button == resetCameraButton) {
			rotationAngle = 0f
			angleOfAttack = 56f
			distance = 10f
		}
		update(Gdx.graphics.deltaTime)
		return false
	}

	override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
		if (button == moveCameraButton) {
			startX = screenX
			startY = screenY
			moveModeOn = false
		}
		update(Gdx.graphics.deltaTime)
		return false
	}

	override fun touchCancelled(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
		// Mirror touchUp so a cancelled gesture cleanly exits move mode instead of crashing.
		return touchUp(screenX, screenY, pointer, button)
	}

	override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
		if (moveModeOn) {
			var deltaX = (screenX - startX).toFloat()
			var deltaY = (startY - screenY).toFloat()
			startX = screenX
			startY = screenY
			if (rotateModeOn) {
				angleOfAttack += deltaY * .25.toFloat()
				rotationAngle += deltaX * .25.toFloat()
			} else {
				if (distance < 150) {
					deltaX *= distance / 100
					deltaY *= distance / 100
				}
				temp.set(target).add(
					sin(rotationAngle) * deltaY + cos(rotationAngle) * -deltaX,
					0f,
					cos(rotationAngle) * deltaY + sin(rotationAngle) * deltaX
				)
			}
			update(Gdx.graphics.deltaTime)
		}
		return false
	}

	override fun scrolled(amountX: Float, amountY: Float): Boolean {
		return if (fastZoomMode) {
			zoom(amountY * translateUnits * 10)
		} else {
			zoom(amountY * translateUnits)
		}
	}

	fun zoom(amount: Float): Boolean {
		distance += amount
		update(Gdx.graphics.deltaTime)
		return true
	}

	override fun keyDown(keycode: Int): Boolean {
		when {
			keycode == rotateMode -> rotateModeOn = true
			keycode == Input.Keys.SHIFT_LEFT -> fastZoomMode = true
		}
		return false
	}

	override fun keyUp(keycode: Int): Boolean {
		if (keycode == rotateMode) {
			rotateModeOn = false
		} else if (keycode == Input.Keys.SHIFT_LEFT) {
			fastZoomMode = false
		}
		return false
	}

	override fun keyTyped(character: Char): Boolean {
		return false
	}

	override fun mouseMoved(screenX: Int, screenY: Int): Boolean {
		return false
	}

	companion object {
		private val temp = Vector3()

		@Deprecated("Normal mapping is no longer controlled by camera input")
		@JvmField
		var yes: Boolean = true

		private fun cos(aoa: Float): Float {
			return kotlin.math.cos(aoa * DEG_TO_RAD.toDouble()).toFloat()
		}

		private fun sin(ang: Float): Float {
			return kotlin.math.sin(ang * DEG_TO_RAD.toDouble()).toFloat()
		}

		/** Polar Projection from Wurst  */
		private fun ppX(x: Float, dist: Float, ang: Float, aoa: Float): Float {
			return (x + dist * kotlin.math.sin(ang * DEG_TO_RAD.toDouble()) * kotlin.math.sin(aoa * DEG_TO_RAD.toDouble())).toFloat()
		}

		private fun ppY(y: Float, dist: Float, ang: Float): Float {
			return (y + dist * kotlin.math.cos(ang * DEG_TO_RAD.toDouble())).toFloat()
		}

		private fun ppZ(z: Float, dist: Float, ang: Float, aoa: Float): Float {
			return (z + dist * kotlin.math.cos(ang * DEG_TO_RAD.toDouble()) * kotlin.math.sin(aoa * DEG_TO_RAD.toDouble())).toFloat()
		}

		private const val DEG_TO_RAD = 0.017453293f
	}

	init {
		update(Gdx.graphics.deltaTime)
	}
}
